{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### frequent_words_with_mismatches(text, k, d)\n",
    "\n",
    "Final Challenge: Find a DnaA box in Salmonella enterica.\n",
    "\n",
    "* Definição\n",
    "O DnaA box é uma sequência conservada localizada na região do origem de replicação (oriC) em bactérias.\n",
    "Esta sequência serve como local de ligação para a proteína DnaA, que é essencial para iniciar a replicação do DNA.\n",
    "\n",
    "* Função do DnaA box\n",
    "- Ligação da proteína DnaA:A proteína DnaA se liga especificamente às sequências do DnaA box.\n",
    "A ligação de múltiplas proteínas DnaA a vários DnaA boxes promove o desenrolamento inicial do DNA na origem de replicação (oriC).\n",
    "\n",
    "- Iniciação da replicação: O desenrolamento do DNA nas proximidades do DnaA box cria uma estrutura de fita simples que permite o recrutamento de outras proteínas de replicação, como a helicase (DnaB) e a primase."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimum_skew(genome):\n",
    "    skew_list = [0]\n",
    "    for i in range(len(genome)):\n",
    "        nucleotide = genome[i]\n",
    "        if nucleotide == 'C':\n",
    "            skew_list.append(skew_list[-1] - 1)\n",
    "        elif nucleotide == 'G':\n",
    "            skew_list.append(skew_list[-1] + 1)\n",
    "        else:\n",
    "            skew_list.append(skew_list[-1])\n",
    "    min_value = min(skew_list)\n",
    "    index_list = [i for i, x in enumerate(skew_list) if x == min_value]\n",
    "    index_list_string = ''\n",
    "    for _ in index_list:\n",
    "        index_list_string += str(_) + ' '\n",
    "    return index_list_string.rstrip()\n",
    "\n",
    "f = open('Salmonella_enterica.txt', 'r')\n",
    "genome_text = f.read().replace('\\n', '')\n",
    "window_start = minimum_skew(genome_text).split()[0]\n",
    "window = genome_text[int(window_start):((int(window_start)) + 500)]\n",
    "\n",
    "def maxmap(worddict):\n",
    "    frequentpatterns = []\n",
    "    max_count = max(worddict.values())\n",
    "    for key, value in worddict.items():\n",
    "        if worddict[key] == max_count:\n",
    "            frequentpatterns.append(key)\n",
    "    return frequentpatterns\n",
    "\n",
    "def hamming_distance(p, q):\n",
    "    list1 = list(p)\n",
    "    list2 = list(q)\n",
    "    count = 0\n",
    "    for i in range(len(list1)):\n",
    "        if list1[i] != list2[i]:\n",
    "            count += 1\n",
    "    return count\n",
    "\n",
    "def suffix(pattern):\n",
    "    suffix_pattern = pattern[1:]\n",
    "    return suffix_pattern\n",
    "\n",
    "def first_symbol(pattern):\n",
    "    first_symbol = pattern[0]\n",
    "    return first_symbol\n",
    "\n",
    "def neighbors(pattern, d):\n",
    "    nucleotides = ['A', 'T', 'G', 'C']\n",
    "    if d == 0:\n",
    "        return pattern\n",
    "    elif len(pattern) == 1:\n",
    "        return nucleotides\n",
    "    neighborhood = []\n",
    "    suffix_neighbors = neighbors(suffix(pattern), d)\n",
    "    for string_text in suffix_neighbors:\n",
    "        if hamming_distance(suffix(pattern), string_text) < d:\n",
    "            for nucleotide in nucleotides:\n",
    "                neighbor = nucleotide + string_text\n",
    "                neighborhood.append(neighbor)\n",
    "        else:\n",
    "            neighbor = first_symbol(pattern) + string_text\n",
    "            neighborhood.append(neighbor)\n",
    "    return neighborhood\n",
    "\n",
    "def reversecomplement(forward_pattern):\n",
    "    large_to_small = {'A':'a', 'T':'t', 'G':'g','C':'c'}\n",
    "    for key, value in large_to_small.items():\n",
    "        if key in forward_pattern:\n",
    "            forward_pattern = forward_pattern.replace(key, value)\n",
    "    complementary = {'a':'T','t':'A','g':'C','c':'G'}\n",
    "    for key, value in complementary.items():\n",
    "        if key in forward_pattern:\n",
    "            forward_pattern = forward_pattern.replace(key, value)\n",
    "    reversecomplementary = forward_pattern[::-1].strip()\n",
    "    return reversecomplementary\n",
    "\n",
    "def frequent_words_with_mismatches(text, k, d):\n",
    "    patterns = []\n",
    "    freqmap = {}\n",
    "    n = len(text)\n",
    "    for i in range(n-k+1):\n",
    "        pattern = text[i:(i+k)]\n",
    "        reverse_pattern = reversecomplement(pattern)\n",
    "        neighborhood = neighbors(pattern, d)\n",
    "        reverse_neighborhood = neighbors(reverse_pattern, d)\n",
    "        for j in range(len(neighborhood)):\n",
    "            neighbor = neighborhood[j]\n",
    "            if neighbor not in freqmap:\n",
    "                freqmap[neighbor] = 1\n",
    "            else:\n",
    "                freqmap[neighbor] += 1\n",
    "        for j in range(len(reverse_neighborhood)):\n",
    "            reverse_neighbor = reverse_neighborhood[j]\n",
    "            if reverse_neighbor not in freqmap:\n",
    "                freqmap[reverse_neighbor] = 1\n",
    "            else:\n",
    "                freqmap[reverse_neighbor] += 1\n",
    "    m = maxmap(freqmap)\n",
    "    for pattern, count in freqmap.items():\n",
    "        if pattern in m:\n",
    "            patterns.append(pattern)\n",
    "    return patterns\n",
    "\n",
    "x = frequent_words_with_mismatches(window, 9, 1)\n",
    "print(x)\n",
    "\n",
    "# new = open('DnaA box frequent kmers.txt', 'w')\n",
    "# for element in x:\n",
    "#     new.write(element + ' ')\n",
    "# new.close()"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
